### PR TITLE
MAUI-4362 Include Notify data error info and data error info in dataform validation

### PR DESCRIPTION
**Pull request by devanathan-vaithiyanathan**
I have include INotifyDataErrorInfo and DataErrorInfo in dataform validation.

### CODE REVIEW

File: MAUI/DataForm/Validation.md [hunk 1]
```diff
@@ -15,6 +15,177 @@ The data form validates the data and user input to update the correct value in t
 
 The supported built in validations are as follows:
 
+#### IDataErrorInfo
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:18]: `IDataErrorInfo` add link

File: MAUI/DataForm/Validation.md [hunk 2]
```diff
@@ -15,6 +15,177 @@ The data form validates the data and user input to update the correct value in t
...
+You can validate the data by implementing the `IDataErrorInfo` interface in the data object class.
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : IDataErrorInfo, INotifyPropertyChanged
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:24]: check INotifyPropertyChanged is required, if not remove related proeprties

File: MAUI/DataForm/Validation.md [hunk 3]
```diff
@@ -15,6 +15,177 @@ The data form validates the data and user input to update the correct value in t
...
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+
+#### INotifyDataErrorInfo
+
+You can validate the data by implementing the `INotifyDataErrorInfo` interface in the data object class. This interface has three members,
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:103]: `INotifyDataErrorInfo` add link

File: MAUI/DataForm/Validation.md [hunk 4]
```diff
@@ -15,6 +15,177 @@ The data form validates the data and user input to update the correct value in t
...
+
+* HasErrors property
+* GetErrors method
+* ErrorsChanged event
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : INotifyDataErrorInfo, INotifyPropertyChanged
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:112]: Check is this required INotifyPropertyChanged

Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:112]: INotifyPropertyChanged removed

File: MAUI/DataForm/Validation.md (hunk 2 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:24]: INotifyPropertyChanged removed and updated in property also

File: MAUI/DataForm/Validation.md (hunk 1 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:18]: link added

File: MAUI/DataForm/Validation.md (hunk 3 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:103]: link added

File: MAUI/DataForm/Validation.md [hunk 5]
```diff
@@ -15,6 +15,115 @@ The data form validates the data and user input to update the correct value in t
...
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : IDataErrorInfo
+{
+    private int employeeID;
+    private string name;
+    private string title;       
+
+    public event PropertyChangedEventHandler PropertyChanged;
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:30]: private int employeeID;
    private string name;
    private string title;       

    public event PropertyChangedEventHandler PropertyChanged;


    public EmployeeInfo()
    {

    }

remove if not needed

File: MAUI/DataForm/Validation.md [hunk 6]
```diff
@@ -15,6 +15,115 @@ The data form validates the data and user input to update the correct value in t
...
+    [Display(AutoGenerateField = false)]
+    public string Error
+    {
+        get
+        {
+            return string.Empty;
+        }
+    }
+
+    public string this[string columnName]
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:52]: [Display(AutoGenerateField = false)]

Restrict auto generation

File: MAUI/DataForm/Validation.md [hunk 7]
```diff
@@ -15,6 +15,115 @@ The data form validates the data and user input to update the correct value in t
...
+* HasErrors property
+* GetErrors method
+* ErrorsChanged event
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : INotifyDataErrorInfo
+{
+    private int employeeID;
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:84]: private int employeeID;
    private string name;
    private string title;

    public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;

    public EmployeeInfo()
    {

    }
remove

Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:84]: private fields removed

File: MAUI/DataForm/Validation.md (hunk 6 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:52]: attribute added

File: MAUI/DataForm/Validation.md (hunk 5 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:30]: private fields removed

File: MAUI/DataForm/Validation.md [hunk 8]
```diff
@@ -15,6 +15,115 @@ The data form validates the data and user input to update the correct value in t
...
+    {
+        get
+        {
+            return false;
+        }
+    }
+
+    public IEnumerable GetErrors(string propertyName)
+    {
+        var list = new List<string>();
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:114]: var list = new List<string>();
        if (propertyName == nameof(Title) && this.Title.Contains("Marketing"))
        {
            list.Add("Marketing is not allowed");
            }
        return list;

File: MAUI/DataForm/Validation.md [hunk 9]
```diff
@@ -15,6 +15,115 @@ The data form validates the data and user input to update the correct value in t
...
+        {
+            return string.Empty;
+        }
+    }
+
+    public string this[string columnName]
+    {
+        get
+        {
+            if (!columnName.Equals("Title"))
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:53]: if (columnName == nameof(Title) && this.Title.Contains("Marketing"))
{
return "Marketing is not allowed";
}
return string.Empty;

File: MAUI/DataForm/Validation.md (hunk 8 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:108]: code updated

File: MAUI/DataForm/Validation.md (hunk 9 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:53]: code changes updated

File: MAUI/DataForm/Validation.md (hunk 1 above)
Comment (karthikraja-arumugam) [reply MAUI/DataForm/Validation.md:18]: move both topics below date range attribute

File: MAUI/DataForm/Validation.md [hunk 10]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+You can validate the data by implementing the [IDataErrorInfo](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.idataerrorinfo?view=net-6.0) interface in the data object class.
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : IDataErrorInfo
+{
+    public event PropertyChangedEventHandler PropertyChanged;
+
+    public EmployeeInfo()
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:28]: public EmployeeInfo()
    {

    }
    
    remove

File: MAUI/DataForm/Validation.md [hunk 11]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
 
+#### IDataErrorInfo
+You can validate the data by implementing the [IDataErrorInfo](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.idataerrorinfo?view=net-6.0) interface in the data object class.
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : IDataErrorInfo
+{
+    public event PropertyChangedEventHandler PropertyChanged;
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:26]: public event PropertyChangedEventHandler PropertyChanged;

Check is this needed, if not remove

File: MAUI/DataForm/Validation.md [hunk 12]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+
+    public EmployeeInfo()
+    {
+
+    }
+
+    public int EmployeeID { get; set; }
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:33]: Use string property

public string EmployeeID { get; set; }

File: MAUI/DataForm/Validation.md [hunk 13]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+
+{% endhighlight %}
+{% endtabs %}
+
+
+#### INotifyDataErrorInfo
+
+You can validate the data by implementing the [INotifyDataErrorInfo](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo?view=net-6.0) interface in the data object class. This interface has three members,
+
+* HasErrors property
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:71]: * HasErrors property
* GetErrors method
* ErrorsChanged event

add links for all three

File: MAUI/DataForm/Validation.md [hunk 14]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+* GetErrors method
+* ErrorsChanged event
+
+{% tabs %}
+{% highlight C# %}
+
+public class EmployeeInfo : INotifyDataErrorInfo
+{
+    public event PropertyChangedEventHandler PropertyChanged;
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:80]: public event PropertyChangedEventHandler PropertyChanged;

Is this required, if not remove

File: MAUI/DataForm/Validation.md [hunk 15]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+    
+    public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
+
+    public EmployeeInfo()
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:84]: public EmployeeInfo()
    {

    }
    
    remove

File: MAUI/DataForm/Validation.md [hunk 16]
```diff
@@ -15,6 +15,108 @@ The data form validates the data and user input to update the correct value in t
...
+    {
+
+    }
+
+    public int EmployeeID { get; set; }
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:89]: public string EmployeeID { get; set; }

File: MAUI/DataForm/Validation.md (hunk 10 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:28]: constructor removed

File: MAUI/DataForm/Validation.md (hunk 11 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:26]: removed

File: MAUI/DataForm/Validation.md (hunk 12 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:33]: int type changed to string

File: MAUI/DataForm/Validation.md (hunk 14 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:80]: removed

File: MAUI/DataForm/Validation.md (hunk 15 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:84]: removed

File: MAUI/DataForm/Validation.md (hunk 16 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:89]: int type to string changed

File: MAUI/DataForm/Validation.md (hunk 13 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:71]: links added

File: MAUI/DataForm/Validation.md [hunk 17]
```diff
@@ -15,6 +15,94 @@ The data form validates the data and user input to update the correct value in t
 
 The supported built in validations are as follows:
 
+#### IDataErrorInfo
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:18]: move both topics below date range attribute

File: MAUI/DataForm/Validation.md [hunk 18]
```diff
@@ -15,6 +15,94 @@ The data form validates the data and user input to update the correct value in t
...
+public class EmployeeInfo : INotifyDataErrorInfo
+{
+    public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
+
+    public string EmployeeID { get; set; }
+
+    public string Name { get; set; }
+
+    public string Title { get; set; }
+
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:80]: remove line

File: MAUI/DataForm/Validation.md [hunk 19]
```diff
@@ -15,6 +15,94 @@ The data form validates the data and user input to update the correct value in t
...
+            return string.Empty;
+        }
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:59]: remove line

File: MAUI/DataForm/Validation.md (hunk 17 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:18]: content moved

File: MAUI/DataForm/Validation.md (hunk 18 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:80]: empty line removed

File: MAUI/DataForm/Validation.md (hunk 19 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:59]: empty line removed

File: MAUI/DataForm/Validation.md [hunk 20]
```diff
@@ -15,6 +15,94 @@ The data form validates the data and user input to update the correct value in t
...
+            }
+
+            return string.Empty;
+        }
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
```
[the remark this answers is not in the document]
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:93]: sample links added

File: MAUI/DataForm/Validation.md [hunk 21]
```diff
@@ -49,6 +49,97 @@ public DateTime JoinDate { get; set; }
 
 Refer [here](https://help.syncfusion.com/maui/dataform/data-annotations#dateformdaterange-attribute) to know more about date range attribute in DataForm.
```
Comment (Jeyasri-Murugan) [inline MAUI/DataForm/Validation.md:51]: Add in over view of Built in validations title

Built in validations
The supported built in validations such as attribute, IDataErrorInfo,  INotifyDataErrorInfo

Add link (attribute, IDataErrorInfo,  INotifyDataErrorInfo

File: MAUI/DataForm/Validation.md [hunk 22]
```diff
@@ -49,6 +49,97 @@ public DateTime JoinDate { get; set; }
...
+
+    [Display(AutoGenerateField = false)]
+    public string this[string columnName]
+    {
+        get
+        {
+            if (columnName == nameof(Title) && this.Title.Contains("Marketing"))
+            {
+                return "Marketing is not allowed";
+            }
```
Comment (Jeyasri-Murugan) [inline MAUI/DataForm/Validation.md:84]: Change 3#
### Validate using the validation attribute
### Validate using the IDataErrorInfo interface
### Validate using the INotifyDataErrorInfo inferface

Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:84]: changed

File: MAUI/DataForm/Validation.md [hunk 23]
```diff
@@ -49,6 +51,97 @@ public DateTime JoinDate { get; set; }
...
+        }
+
+        return list;
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+You can download the entire source code [here](https://github.com/SyncfusionExamples/maui-dataform/tree/master/NotifyDataErrorInfoSample)
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:143]: check sample reference content and follow same pattern.
https://help.syncfusion.com/maui/dataform/grouping#changing-the-padding-for-group-headers-and-editors

File: MAUI/DataForm/Validation.md [hunk 24]
```diff
@@ -49,6 +51,97 @@ public DateTime JoinDate { get; set; }
...
+
+            return string.Empty;
+        }
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+You can download the entire source code [here](https://github.com/SyncfusionExamples/maui-dataform/tree/master/DataErrorInfoSample)
```
Comment (karthikraja-arumugam) [inline MAUI/DataForm/Validation.md:96]: check sample reference content and follow same pattern.
https://help.syncfusion.com/maui/dataform/grouping#changing-the-padding-for-group-headers-and-editors

Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:96]: pattern updated

File: MAUI/DataForm/Validation.md (hunk 23 above)
Comment (devanathan-vaithiyanathan) [reply MAUI/DataForm/Validation.md:143]: pattern updated

**Review by florenceliliana — COMMENTED** [review]
The content is fine.

**Review by Jeyasri-Murugan — APPROVED** [review]

**Pull request merged**